example/methods/finite_difference/finite_difference.cpp

```diff
@@ -0,0 +1,194 @@
+/****************************************************************************
+ * Copyright (c) 2018-2022 by the Cabana authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the Cabana library. Cabana is distributed under a   *
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <array>
+#include <math.h>
+
+#include <mpi.h>
+
+#include <Kokkos_Core.hpp>
+
+#include <Cajita.hpp>
+
+// Apply boundary conditions. For simplicity of the code, rebuild the index
+// spaces each step even though it's a fixed geometry. Apply all boundaries in
+// one kernel for performance.
+template <class ExecutionSpace, class LocalGridType, class ArrayType>
+void updateBoundaries( ExecutionSpace exec_space, LocalGridType local_grid,
+                       ArrayType& field )
+{
+    // Create boundary indices for each plane.
+    Kokkos::Array<Cajita::IndexSpace<3>, 6> boundary_spaces;
+    // Store the boundary details in device-accessible fixed-size arrays.
+    Kokkos::Array<Kokkos::Array<int, 3>, 6> planes;
+    // Generate the boundary condition index spaces.
+    int count = 0;
+    for ( int d = 0; d < 3; d++ )
+    {
+        for ( int dir = -1; dir < 2; dir += 2 )
+        {
+            planes[count] = { 0, 0, 0 };
+            planes[count][d] = dir;
+
+            // Get the boundary indices for this plane (each one is a separate,
+            // contiguous index space).
+            boundary_spaces[count] = local_grid->boundaryIndexSpace(
+                Cajita::Own(), Cajita::Cell(), planes[count][0],
+                planes[count][1], planes[count][2] );
+            count++;
+        }
+    }
+
+    // Update the boundary on each face of the cube. Pass the vector of index
+    // spaces to avoid launching 6 separate kernels.
+    Cajita::grid_parallel_for(
+        "boundary_update", exec_space, boundary_spaces,
+        KOKKOS_LAMBDA( const int b, const int i, const int j, const int k ) {
+            // Neumann boundary condition example
+            field( i, j, k, 0 ) = field( i - planes[b][0], j - planes[b][1],
+                                         k - planes[b][2], 0 );
+        } );
+}
+
```

Comment:
It might be a good idea to add comment about the problem. I think this is about unsteady heat equation problem in 3d.

The governing equstion :  \rho c_p dT/dt = k \nabla{T} + q ( heat source ? )
Neumann boundary condition (describing feat flux ) : k \nabla{T} \cdot n = f

Reply:
@colemanjs can you add a more detailed comment below?